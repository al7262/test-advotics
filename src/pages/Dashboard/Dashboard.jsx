/* eslint-disable react/forbid-prop-types */
import React from 'react';
import PropTypes from 'prop-types';
import { useSelector } from 'react-redux';
import PeriodSelection from '../../components/elements/PeriodSelection';
import Card from '../../components/elements/Card';
import Chart from '../../components/elements/Chart';
import HelpIcon from '../../assets/help.png';
import DownArrow from '../../assets/down-arrow.svg';
import CartIcon from '../../assets/cart.svg';
import { currency, thousandSeparator } from '../../utils/format';
import styles from './index.css';

export default function Dashboard() {
  const { bestSelling, topCompetitor } = useSelector((s) => s.dashboard);
  return (
    <section className={styles.root}>
      <header className={styles.header}>
        <h1>Dashboard</h1>
        <PeriodSelection />
      </header>
      <div className={styles.snackbar}>
        <h2>Market Insight</h2>
        <img src={HelpIcon} alt="help-icon-advotics" />
        <h6>Click Here for Help</h6>
        <span />
      </div>
      <article className={styles.content}>
        <Card className={styles.sales} title="Sales Turnover" size="small">
          <b>{currency(3600000)}</b>
          <small>
            <img src={DownArrow} alt="down-arrow-advotics" />
            <b>13.8% </b>
            last period in products sold
          </small>
          <img src={CartIcon} alt="cart-icon-advotics" />
        </Card>
        <Card className={styles.chart} title="Average Purchase Value" fixed>
          <Chart />
        </Card>
        <Card className={styles.chart} title="Best Selling SKU" fixed>
          <ProductList data={bestSelling} />
        </Card>
        <Card className={styles.chart} title="Top Competitor SKU" fixed>
          <ProductList data={topCompetitor} />
        </Card>
      </article>
    </section>
  );
}

const ProductList = ({ data }) => {
  const slicedData = data.slice(0, 5);
  return (
    <div className={styles['product-list']}>
      {slicedData.map((item, key) => (
        <div className={styles['product-item']} key={`produk-${key + 1}`}>
          <img src={item.image} alt={`product-${item.name}`} />
          <h5 title={item.name}>{item.name}</h5>
          <p title={currency(item.price)}>{currency(item.price)}</p>
          <p title={`${thousandSeparator(item.quantity)} pcs`}>{`${thousandSeparator(item.quantity)} pcs`}</p>
        </div>
      ))}
    </div>
  );
};

ProductList.defaultProps = {
  data: [],
};

ProductList.propTypes = {
  data: PropTypes.array,
};
