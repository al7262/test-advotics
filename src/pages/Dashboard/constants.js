export const LOADING = 'Dashboard/loading';
export const DATA_FETCHED = 'Dashboard/data-fetched';
