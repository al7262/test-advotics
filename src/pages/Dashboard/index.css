.header {
    display: flex;
    gap: 1rem;
    align-items: center;
    padding-bottom: 1.875rem;
}

.header > h1 {
    color: var(--black-color-3);
}

.snackbar {
    height: 3rem;
    padding: .75rem 1rem;
    background-color: var(--primary-color);
    color: var(--white-color) !important;
    display: flex;
    align-items: center;
    margin-bottom: 1rem;
}

.snackbar > img {
    height: 1.25rem;
    margin-left: auto;
    margin-right: .25rem;
}

.snackbar > h6 {
    text-decoration: underline;
    margin-right: 1rem;
}

.snackbar > span {
    border: solid var(--white-color);
    border-width: 0 .125rem .125rem 0;
    display: inline-block;
    padding: .25rem;
    transform: rotate(-135deg);
    margin-top: .5rem;
}

.content {
    display: grid;
    grid-template-columns: 2fr 1fr 1fr;
    gap: 1rem;
}

.sales {
    width: 17.75rem;
    grid-column: 1 / span 3;
}

.sales > main {
    display: grid;
    grid-template-columns: auto auto;
    column-gap: 1rem;
}

.sales > main > b {
    font-size: 1.5rem;
}

.sales > main > img {
    width: 3rem;
    margin-right: auto;
    grid-row: 1 / span 2;
    grid-column: 2;
}

.sales > main > small {
    height: .75rem;
    display: flex;
    align-items: center;
    gap: .25rem;
}

.sales > main > small > img {
    height: 75%;
}

.sales > main > small > b {
    font-weight: 600;
    color: var(--red-color);
}

.product-list {
    width: 100%;
}

.product-item {
    height: 3.75rem;
    width: 100%;
    border: 0.0625rem solid var(--black-color-5);
    border-radius: .25rem;
    display: grid;
    gap: .5rem;
    grid-template-columns: 3.75rem auto auto;
    grid-template-rows: auto auto;
    margin-bottom: .5rem;
}

.product-item > img {
    grid-row: 1 / span 2;
    height: 100%;
    width: 100%;
    display: block;
    object-fit: cover;
}

.product-item > h5 {
    grid-column: 2 / span 2;
    margin-top: auto;
    padding-right: 1.5rem;
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
}

.product-item > p {
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
}

.product-item > p:last-of-type {
    margin-left: auto;
    padding-right: 1.5rem;
}

.product-item:first-child {
    height: 5rem;
    grid-template-columns: 5rem auto auto;
    position: relative;
    z-index: 1;
}

.product-item:first-child::before {
    content: "";
    position: absolute;
    top: 0;
    left: 0;
    height: 100%;
    width: 100%;
    border-radius: .25rem;
    background-color: var(--secondary-color);
    opacity: .5;
    z-index: -1;
}

.product-item:first-child > h5 {
    font-size: 1.25rem;
}

.product-item:first-child > p {
    font-size: .875rem;
}