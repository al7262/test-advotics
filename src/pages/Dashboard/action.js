/* eslint-disable import/prefer-default-export */
import { LOADING, DATA_FETCHED } from './constants';

function loadingAction(isLoading) {
  return { type: LOADING, isLoading };
}

function dataFetchedAction(data) {
  return { type: DATA_FETCHED, data };
}

export function fetchData(payload) {
  return (dispatch) => {
    dispatch(loadingAction(true));
    try {
      const data = payload();
      dataFetchedAction(data);
    } catch (err) {
      console.log(err);
    }
  };
}
