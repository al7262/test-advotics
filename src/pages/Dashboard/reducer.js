import { LOADING, DATA_FETCHED } from './constants';
import ProductImg from '../../assets/product2.webp';
import { sortByDescending } from '../../utils/format';

const bestSelling = sortByDescending([0, 1, 2, 3, 4].map((n) => ({
  name: `Produk ${n}`,
  id: n,
  image: ProductImg,
  price: 100000,
  quantity: Math.floor(Math.random() * 100),
})), 'quantity');

const topCompetitor = [0, 1, 2, 3, 4].map((n) => ({
  name: `Produk ${n}`,
  id: n,
  image: ProductImg,
  price: 100000,
  quantity: Math.floor(Math.random() * 100),
}));

const initialState = {
  averagePurchase: [],
  bestSelling,
  topCompetitor,
  isLoading: false,
};

export default function reducer(state = initialState, action = {}) {
  const {
    data, isLoading, type,
  } = action;

  switch (type) {
    case LOADING:
      return {
        ...state,
        isLoading,
      };
    case DATA_FETCHED:
      return {
        ...state,
        ...data,
        isLoading: false,
      };
    default:
      return state;
  }
}
