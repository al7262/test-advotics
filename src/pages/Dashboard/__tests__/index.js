/* eslint-disable react/jsx-filename-extension */
/* eslint-disable no-undef */
import React from 'react';
import { useSelector } from 'react-redux';
import ShallowRenderer from 'react-test-renderer/shallow';
import Dashboard from '../Dashboard';

useSelector.mockImplementation((fn) => {
  fn({});
  return ({ bestSelling: [], topCompetitor: [] });
});

describe('src/pages/Dashboard', () => {
  test('render', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<Dashboard />);
    expect(tree).toMatchSnapshot();
  });
});
