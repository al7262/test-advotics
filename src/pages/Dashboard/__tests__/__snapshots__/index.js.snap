// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/pages/Dashboard render 1`] = `
<section>
  <header>
    <h1>
      Dashboard
    </h1>
    <PeriodSelection />
  </header>
  <div>
    <h2>
      Market Insight
    </h2>
    <img
      alt="help-icon-advotics"
      src={
        Object {
          "process": [Function],
        }
      }
    />
    <h6>
      Click Here for Help
    </h6>
    <span />
  </div>
  <article>
    <Card
      className=""
      fixed={false}
      size="small"
      title="Sales Turnover"
    >
      <b>
        Rp 3,600,000
      </b>
      <small>
        <img
          alt="down-arrow-advotics"
          src={
            Object {
              "process": [Function],
            }
          }
        />
        <b>
          13.8% 
        </b>
        last period in products sold
      </small>
      <img
        alt="cart-icon-advotics"
        src={
          Object {
            "process": [Function],
          }
        }
      />
    </Card>
    <Card
      className=""
      fixed={true}
      size="normal"
      title="Average Purchase Value"
    >
      <Chart />
    </Card>
    <Card
      className=""
      fixed={true}
      size="normal"
      title="Best Selling SKU"
    >
      <ProductList
        data={Array []}
      />
    </Card>
    <Card
      className=""
      fixed={true}
      size="normal"
      title="Top Competitor SKU"
    >
      <ProductList
        data={Array []}
      />
    </Card>
  </article>
</section>
`;
